File: libraries/react-native/tests/core/account-token.test.ts
import { describe, expect, it } from "vitest";

import {
  deriveAccountToken,
  uuidV5,
  VOIDHASH_ACCOUNT_TOKEN_NAMESPACE,
} from "../../src/core/utils/account-token";

const DNS_NAMESPACE = "6ba7b810-9dad-11d1-80b4-00c04fd430c8";

const vectors = [
  ["user-123", "3501e751-7582-58f9-9c1d-533c7466049f"],
  ["USER-123", "c6eb5cb5-739d-52a9-9a32-6a0fb4be71cc"],
  ["vh:anon:k2j4h5g6f7", "22c4cde2-2c40-5266-9699-35f7d271e1a8"],
  ["naïve@exämple.com", "84080bfc-8e5a-5daf-9ec4-c2221ea8d948"],
] as const;

describe("account token derivation", () => {
  it("reproduces the RFC 4122 UUIDv5 sanity anchor", () => {
    expect(uuidV5(DNS_NAMESPACE, "www.example.com")).toBe("2ed6657d-e927-568b-95e1-2665a8aea6a2");
  });

  it("pins the shared namespace", () => {
    expect(uuidV5(DNS_NAMESPACE, "appaccounttoken.voidhash.com")).toBe(
      VOIDHASH_ACCOUNT_TOKEN_NAMESPACE,
    );
  });

  for (const [distinctId, expected] of vectors) {
    it(`matches the backend vector for ${JSON.stringify(distinctId)}`, () => {
      expect(deriveAccountToken(distinctId)).toBe(expected);
    });
  }

  it("is deterministic and lowercase for long inputs", () => {
    const distinctId = "person:".repeat(200);
    const first = deriveAccountToken(distinctId);

    expect(deriveAccountToken(distinctId)).toBe(first);
    expect(first).toMatch(/^[0-9a-f]{8}-[0-9a-f]{4}-5[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/);
  });
});

File: libraries/android/sdk/src/test/java/com/voidhash/sdk/identity/AccountTokenTest.kt
package com.voidhash.sdk.identity

import com.voidhash.sdk.cache.CacheManager
import com.voidhash.sdk.cache.InMemoryCacheAdapter
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

private const val DNS_NAMESPACE = "6ba7b810-9dad-11d1-80b4-00c04fd430c8"

class AccountTokenTest {
    @Test
    fun `derives the pinned RFC 4122 vector`() {
        assertEquals(
            "2ed6657d-e927-568b-95e1-2665a8aea6a2",
            AccountToken.uuidV5(DNS_NAMESPACE, "www.example.com"),
        )
    }

    @Test
    fun `the voidhash namespace is derived from the DNS namespace`() {
        assertEquals(
            AccountToken.NAMESPACE,
            AccountToken.uuidV5(DNS_NAMESPACE, "appaccounttoken.voidhash.com"),
        )
    }

    @Test
    fun `derives the pinned distinct id vectors`() {
        assertEquals("3501e751-7582-58f9-9c1d-533c7466049f", AccountToken.derive("user-123"))
        assertEquals("c6eb5cb5-739d-52a9-9a32-6a0fb4be71cc", AccountToken.derive("USER-123"))
        assertEquals("22c4cde2-2c40-5266-9699-35f7d271e1a8", AccountToken.derive("vh:anon:k2j4h5g6f7"))
    }

    @Test
    fun `hashes UTF-8 bytes, not code units`() {
        assertEquals(
            "84080bfc-8e5a-5daf-9ec4-c2221ea8d948",
            AccountToken.derive("naïve@exämple.com"),
        )
    }
}

class IdentityStoreTest {
    private fun store(anonymousId: String = "vh:anon:test") =
        IdentityStore(CacheManager(InMemoryCacheAdapter())) { anonymousId }

    @Test
    fun `generates and persists an anonymous distinct id`() {
        val identityStore = store()
        val distinctId = identityStore.getDistinctId()

        assertTrue(distinctId.startsWith("vh:anon:"))
        assertEquals(distinctId, identityStore.getDistinctId())
        assertTrue(identityStore.isAnonymous())
    }

    @Test
    fun `identify replaces the distinct id and reset drops it`() {
        val identityStore = store()
        identityStore.getDistinctId()

        identityStore.setDistinctId("user-123")
        assertEquals("user-123", identityStore.getDistinctId())
        assertTrue(!identityStore.isAnonymous())

        identityStore.reset()
        assertEquals("vh:anon:test", identityStore.getDistinctId())
    }
}
